File: ios/Plugin/Plugin/StorageDatabaseHelper.swift
//
//  StorageDatabaseHelper.swift
//  Plugin
//
//  Created by  Quéau Jean Pierre on 02/05/2018.
//  Copyright © 2018 Max Lynch. All rights reserved.
//

import Foundation
import SQLite
class StorageDatabaseHelper {
    let DATABASE_NAME: String = "storageSQLite.db"
    let TABLE_INDEX: Table = Table("sqlite_sequence")
    let IDX_COL_NAME: Expression<String> = Expression<String>("name")
    let IDX_COL_SEQ: Expression<Int64> = Expression<Int64>("seq")
    let TABLE_STORAGE_NAME = "storage_table"
    let TABLE_STORAGE: Table = Table("storage_table")
    let COL_ID: Expression<Int64> = Expression<Int64>("id")
    let COL_NAME: Expression<String> = Expression<String>("name")
    let COL_VALUE: Expression<String> = Expression<String>("value")
    // define the path for the database
    let path: String = NSSearchPathForDirectoriesInDomains(
        .documentDirectory, .userDomainMask, true
        ).first!
    
    init() {
        print("path: \(path)")
        // connect to the database (create if doesn't exist)
        guard let db = try? Connection("\(path)/\(DATABASE_NAME)") else {
            let error:String = "init: Error Database connection failed"
            print(error)
            return
        }
        // create table
        do {
            try db.run(TABLE_STORAGE.create(ifNotExists: true) { t in
                t.column(COL_ID, primaryKey: .autoincrement)
                t.column(COL_NAME,unique: true)
                t.column(COL_VALUE)
            })
            // index COL_NAME
            do {
                try db.run(TABLE_STORAGE.createIndex(COL_NAME, ifNotExists: true))
                
            } catch let error {
                print("init: Error Index creation failed: \(error)")
            }
        } catch let error {
            print("init: Error Table creation failed: \(error)")
        }
    }
    
    func getWritableDatabase() -> Connection? {
        guard let db = try? Connection("\(path)/\(DATABASE_NAME)") else {return nil}
        return db
    }
    
    func getReadableDatabase() -> Connection? {
        guard let db = try? Connection("\(path)/\(DATABASE_NAME)", readonly: true) else {return nil}
        return db
    }
    
    func addData(data:Data) -> Bool {
        var ret: Bool = false
        guard let db: Connection = try? getWritableDatabase()! else {return false}
        // check if the data already exists
        if (try! getDataByName(name: data.name!))!.id != nil {
            ret = updateData(data: data)
        } else {
            do {
                try db.run(TABLE_STORAGE.insert(
                    COL_NAME <- data.name!,
                    COL_VALUE <- data.value!
                ))
                ret = true
            } catch let error {
                print("addData: Error Data insertion failed: \(error)")
            }
        }
        return ret
    }
    
    func updateData(data:Data) -> Bool {
        var ret: Bool = false
        guard let db: Connection = try? getWritableDatabase()! else {return false}
        let mFilter = TABLE_STORAGE.filter(COL_NAME == data.name!)
        do {
            if try db.run(mFilter.update(COL_VALUE <- data.value!)) > 0 {
                ret = true
            } else {
                print("updateData: Error \(COL_NAME) not found")
            }
        } catch let error {
            print("updateData: Error update failed: \(error)")
        }
        return ret
    }
    
    func deleteDataByName(name:String) -> Bool {
        var ret: Bool = false
        guard let db: Connection = try? getWritableDatabase()! else {return false}
        let mFilter = TABLE_STORAGE.filter(COL_NAME == name)
        do {
            if try db.run(mFilter.delete()) > 0 {
                ret = true
            } else {
                print("deleteDataByName: Error \(COL_NAME) not found")
            }
        } catch let error {
            print("deleteDataByName: Error delete failed: \(error)")
        }
        return ret
    }
    
    func deleteAllData() -> Bool {
        var ret: Bool = false
        guard let db: Connection = try? getWritableDatabase()! else {return false}
        do {
            try db.run(TABLE_STORAGE.delete())
            ret = resetIndex()
        } catch let error {
            print("deleteAllData: Error All data delete failed: \(error)")
        }
        return ret
    }
    
    func resetIndex() -> Bool {
        var ret: Bool = false
        guard let db: Connection = try? getWritableDatabase()! else {return false}
        let idxFilter = TABLE_INDEX.filter(IDX_COL_NAME == TABLE_STORAGE_NAME).limit(1)
        do {
            if try db.run(idxFilter.update(IDX_COL_SEQ <- 0)) > 0{
                ret = true
            } else {
                print("resetIndex: Error did not update the index" )
            }
        } catch let error {
            print("resetIndex: Error Index update failed: \(error)")
        }
        return ret
    }
    
    func getDataByName(name:String) -> Data? {
        var retData: Data = Data()
        guard let db: Connection = try? getReadableDatabase()! else {return nil}
        let query = TABLE_STORAGE.filter(COL_NAME == name).limit(1)
        guard let rData = try? db.prepare(query) else {return nil}
        for data in rData {
            retData.id = data[COL_ID]
            retData.name = data[COL_NAME]
            retData.value = data[COL_VALUE]
        }
        return retData
    }
    
    func getAllData() -> Array<Data>? {
        var retArray: Array<Data> = Array<Data>()
        guard let db: Connection = try? getReadableDatabase()! else {return nil}
        guard let retData: AnySequence<Row> = try? db.prepare(TABLE_STORAGE) else { return nil}
        for rData in retData {
            var data = Data()
            data.id = rData[COL_ID]
            data.name = rData[COL_NAME]
            data.value = rData[COL_VALUE]
            retArray.append(data)
        }
        return retArray
    }
}

File: android/src/main/java/com/jeep/plugin/capacitor/capacitordatastoragesqlite/StorageDatabaseHelper.java
//
//  StorageDatabaseHelper.java
//  Plugin
//
//  Created by  Quéau Jean Pierre on 04/05/2018.
//  Copyright © 2018 Max Lynch. All rights reserved.
//
package com.jeep.plugin.capacitor.capacitordatastoragesqlite;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;


public class StorageDatabaseHelper extends SQLiteOpenHelper {
    private static final String TAG = "StorageDatabaseHelper";
    private static StorageDatabaseHelper sInstance;
    // Database Info
//    private static final String DATABASE_NAME = "storageSQLite.db";
//    private static final int DATABASE_VERSION = 1;

    // Table Names
//    private static final String TABLE_STORAGE = "storage_table";
    private String tableName;
    private final String dbName;
    private final int dbVersion;

    // Storage Table Columns
    private static final String COL_ID = "id";
    private static final String COL_NAME = "name";
    private static final String COL_VALUE = "value";
/*
    public static synchronized StorageDatabaseHelper getInstance(Context context, String _dbName, String _tableName, int _vNumber) {
        // Use the application context, which will ensure that you
        // don't accidentally leak an Activity's context.
        // See this article for more information: http://bit.ly/6LRzfx
        if (sInstance == null) {
            sInstance = new StorageDatabaseHelper(context.getApplicationContext(),_dbName,_tableName,_vNumber);
        }
        return sInstance;
    }
*/
    /**
     * Constructor should be private to prevent direct instantiation.
     * Make a call to the static method "getInstance()" instead.
     */
    //private StorageDatabaseHelper(Context context,String _dbName, String _tableName,int _vNumber) {
    StorageDatabaseHelper(Context context, String _dbName, String _tableName, int _vNumber) {
        super(context, _dbName, null, _vNumber);
        dbName = _dbName;
        tableName = _tableName;
        dbVersion = _vNumber;
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        Log.d(TAG, "onCreate: "+ tableName);
        createTable(db,tableName);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        if (oldVersion != newVersion) {
            dropAllTables(db);
            onCreate(db);
        }
    }
    public boolean setTable(String _tableName) {
        SQLiteDatabase db = getWritableDatabase();
        boolean ret = createTable(db,_tableName);
        Log.d(TAG, "setTable: "+ tableName);
        return ret;

    }
    // Insert a data into the database
    public boolean set(Data data) {
        boolean ret = false;
        // Create and/or open the database for writing
        SQLiteDatabase db = getWritableDatabase();
        // Check if data.name does not exist otherwise update it
        Data res = this.get(data.name);
        if (res.id != null) {
            // exists so update it
            return this.update(data);

        } else {
            // does not exist add it
            // wrap our insert in a transaction.
            db.beginTransaction();
            try {

                ContentValues values = new ContentValues();
                values.put(COL_NAME, data.name);
                values.put(COL_VALUE, data.value);

                // do not need to specify the primary key. SQLite auto increments the primary key column.
                db.insertOrThrow(tableName, null, values);
                db.setTransactionSuccessful();
                ret = true;
            } catch (Exception e) {
                Log.d(TAG, "set: Error while trying to add data to database");
            } finally {
                db.endTransaction();
                return ret;
            }
        }
    }

    // update data value into the database
    public boolean update(Data data) {
        boolean ret = false;
        SQLiteDatabase db = getWritableDatabase();
        ContentValues values = new ContentValues();
        values.put(COL_VALUE, data.value);

        // wrap our update in a transaction.
        db.beginTransaction();
        try {
            db.update(tableName, values, COL_NAME + " = ?",
                    new String[] { String.valueOf(data.name) });
            db.setTransactionSuccessful();
            ret = true;
        } catch (Exception e) {
            Log.d(TAG, "update: Error while trying to update " + data.name);
        } finally {
            db.endTransaction();
            return ret;
        }

    }

    // delete a data into the database
    public boolean remove(String name) {
        boolean ret = false;
        SQLiteDatabase db = getWritableDatabase();
        // wrap our delete in a transaction.
        db.beginTransaction();
        try {
            db.delete(tableName, COL_NAME + "= '" + name +"'",
                    null);
            db.setTransactionSuccessful();
            ret = true;
        } catch (Exception e) {
            Log.d(TAG, "remove: Error while trying to delete " + name);
        } finally {
            db.endTransaction();
            return ret;
        }
    }

    // delete a all data into the database
    public boolean clear() {
        boolean ret = false;
        SQLiteDatabase db = getWritableDatabase();
        // wrap our delete in a transaction.
        ContentValues values = new ContentValues();
        values.put("SEQ", 0);
        db.beginTransaction();
        try {
            db.delete(tableName, null, null);
            // set back the key primary index to 0
            db.update("SQLITE_SEQUENCE", values, "NAME = ?",
                    new String[] { String.valueOf(tableName) });
            db.setTransactionSuccessful();
            ret = true;
        } catch (Exception e) {
            Log.d(TAG, "Clear: Error while trying to delete all data");
        } finally {
            db.endTransaction();
            return ret;
        }
    }

    // isKey exists
    public boolean iskey(String name) {
        boolean ret = false;
        Data data = get(name);
        if(data.id != null) ret =true;
        return ret;
    }

    // get a Data
    public Data get(String name) {
        Data data = null;

        String DATA_SELECT_QUERY = "SELECT * FROM "+ tableName +
                " WHERE " + COL_NAME + " = '" + name + "'";
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery(DATA_SELECT_QUERY, null);
        if (cursor.getCount() > 0) {
            try {
                if (cursor.moveToFirst()) {
                    data = new Data();
                    data.id = cursor.getLong(cursor.getColumnIndex(COL_ID));
                    data.name = cursor.getString(cursor.getColumnIndex(COL_NAME));
                    data.value = cursor.getString(cursor.getColumnIndex(COL_VALUE));
                }
            } catch (Exception e) {
                Log.d(TAG, "get: Error while trying to get data from storage database");
            } finally {
                if (cursor != null && !cursor.isClosed()) {
                    cursor.close();
                }
            }
        } else {
            data = new Data();
            data.id = null;
            if (cursor != null && !cursor.isClosed()) {
                cursor.close();
            }
        }
        return data;
    }

    // get All Keys
    public List<String> keys() {
        List<String> data = new ArrayList<>();

        String DATA_SELECT_QUERY = "SELECT * FROM "+ tableName;
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery(DATA_SELECT_QUERY, null);
        if (cursor.getCount() > 0) {
            try {
                if (cursor.moveToFirst()) {
                    do {

                        String key = cursor.getString(cursor.getColumnIndex(COL_NAME));
                        data.add(key);
                    } while(cursor.moveToNext());
                }
            } catch (Exception e) {
                Log.d(TAG, "keys: Error while trying to get all keys from storage database");
            } finally {
                if (cursor != null && !cursor.isClosed()) {
                    cursor.close();
                }
            }
        } else {
            data = Collections.EMPTY_LIST;
            if (cursor != null && !cursor.isClosed()) {
                cursor.close();
            }
        }
        return data;
    }

    // get All Values
    public List<String> values() {
        List<String> data = new ArrayList<>();

        String DATA_SELECT_QUERY = "SELECT * FROM "+ tableName;
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery(DATA_SELECT_QUERY, null);
        if (cursor.getCount() > 0) {
            try {
                if (cursor.moveToFirst()) {
                    do {

                        String value = cursor.getString(cursor.getColumnIndex(COL_VALUE));
                        data.add(value);
                    } while(cursor.moveToNext());
                }
            } catch (Exception e) {
                Log.d(TAG, "values: Error while trying to get all values from storage database");
            } finally {
                if (cursor != null && !cursor.isClosed()) {
                    cursor.close();
                }
            }
        } else {
            data = Collections.EMPTY_LIST;
            if (cursor != null && !cursor.isClosed()) {
                cursor.close();
            }
        }
        return data;
    }

    // get All Data
    public List<Data> keysvalues() {
        List<Data> data = new ArrayList<>();

        String DATA_SELECT_QUERY = "SELECT * FROM "+ tableName;
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery(DATA_SELECT_QUERY, null);
        if (cursor.getCount() > 0) {
            try {
                if (cursor.moveToFirst()) {
                    do {
                        Data newData = new Data();
                        newData.name = cursor.getString(cursor.getColumnIndex(COL_NAME));
                        newData.value = cursor.getString(cursor.getColumnIndex(COL_VALUE));
                        data.add(newData);
                    } while (cursor.moveToNext());
                }
            } catch (Exception e) {
                Log.d(TAG, "keysvalues: Error while trying to get all keys/values from storage database");
            } finally {
                if (cursor != null && !cursor.isClosed()) {
                    cursor.close();
                }
            }
        } else {
            data = Collections.EMPTY_LIST;
            if (cursor != null && !cursor.isClosed()) {
                cursor.close();
            }
        }
        return data;
    }
    private boolean dropAllTables(SQLiteDatabase db) {
        boolean ret = false;
        List<String> tables = new ArrayList<String>();
        Cursor cursor = db.rawQuery("SELECT * FROM sqlite_master WHERE type='table';", null);
        cursor.moveToFirst();
        while (!cursor.isAfterLast()) {
            String tableName = cursor.getString(1);
            if (!tableName.equals("android_metadata") &&
                    !tableName.equals("sqlite_sequence"))
                tables.add(tableName);
            cursor.moveToNext();
        }
        cursor.close();
        try {
            for(String tableName:tables) {
                db.execSQL("DROP TABLE IF EXISTS " + tableName);
            }
        } catch (Exception e) {
            Log.d(TAG, "Error: dropAllTables failed: ",e);
        } finally {
            ret = true;
        }
        return ret;
    }
    private boolean createTable(SQLiteDatabase db, String _tableName) {

        String CREATE_STORAGE_TABLE = "CREATE TABLE IF NOT EXISTS " + _tableName +
                "(" +
                COL_ID + " INTEGER PRIMARY KEY AUTOINCREMENT," + // Define a primary key
                COL_NAME + " TEXT," +
                COL_VALUE + " TEXT" +
                ")";
        String CREATE_INDEX_NAME = "CREATE INDEX IF NOT EXISTS idx_"+_tableName+"_name on " + _tableName +
                " (" + COL_NAME  + ")";
        Log.d(TAG, "onsetTable: "+ _tableName);
        db.beginTransaction();
        try {
            db.execSQL(CREATE_STORAGE_TABLE);
            db.execSQL(CREATE_INDEX_NAME);
            Log.d(TAG, "onCreate: name: database created");
            db.setTransactionSuccessful();
        } catch (Exception e) {
            Log.d(TAG, "set: Error while trying to add a table to database");
            return false;
        } finally {
            db.endTransaction();
            tableName = _tableName;
            return true;
        }
    }
}
